File: 2021/day02/main.pas
program DAY02;

type
    splitResult = array of string;

var
    result: splitResult;
    depth, horizontal, x: integer;
    data: array[0..5] OF STRING = ('forward 5','down 5','forward 8','up 3','down 8','forward 2');

function isSeparator(const ch: char; const separator: string): boolean;
var sep: char;
begin
    for sep in separator do begin
        if sep=ch then begin
            isSeparator := true;
            exit;
        end;
    end;
    isSeparator := false;
end;

function splitString(const str; separator: string): splitResult;
var
    charIndex, itemIndex, len, separatorCount, start: integer;
begin
    len := Length(str);
    if len=0 then begin
        result := nil;
        exit;
    end;
    separatorCount := 0;
    for charIndex := 1 to len do begin
        if isSeparator(str[charIndex], separator) then begin
            inc(separatorCount)
        end;
    end;
    SetLength(result, separatorCount+1);
    itemIndex := 0;
    start := 1;
    for charIndex := 1 to len do begin
        if isSeparator(str[charIndex], separator) then begin
            if charIndex > start then begin
                result[itemIndex] := Copy(str, start, charIndex - start);
                inc(itemIndex);
            end;
            start := charIndex + 1
        end;
    end;
    if len > start then begin
        result[itemIndex] := Copy(str, start, len - start + 1);
        inc(itemIndex);
    end;
    SetLength(result, itemIndex);
end;

begin
    writeln ('');
end.

File: 2021/day02/main.py
from operator import mul

with open("2021/day02/input.txt") as f:
    DATA = [(move, int(value)) for move, value in map(str.split, f)]


def part_one():
    return mul(
        sum(x[1] if x[0] == "down" else -x[1] for x in DATA if x[0] in ("up, down")),
        sum(x[1] for x in DATA if x[0] == "forward"),
    )


def part_two():
    aim = depth = horizontal = 0
    for x in DATA:
        if x[0] == "forward":
            horizontal += x[1]
            depth += x[1] * aim
        else:
            aim += -x[1] if x[0] == "up" else x[1]
    return depth * horizontal


print(part_one())
print(part_two())
